core/src/transport/map_err.rs

```diff
@@ -92,23 +78,177 @@ where
 
 impl<T, F> MuxedTransport for MapErr<T, F>
 where
-    T: MuxedTransport + 'static,                     // TODO: 'static :-/
-    F: FnOnce(IoError) -> IoError + Clone + 'static, // TODO: 'static :-/
+    T: MuxedTransport,
+    F: FnOnce(IoError) -> IoError + Clone,
 {
-    type Incoming = Box<Future<Item = Self::IncomingUpgrade, Error = IoError>>;
-    type IncomingUpgrade =
-        Box<Future<Item = (Self::Output, Self::MultiaddrFuture), Error = IoError>>;
+    type Incoming = MapErrIncoming<T, F>;
+    type IncomingUpgrade = MapErrIncomingUpgrade<T, F>;
 
+    #[inline]
     fn next_incoming(self) -> Self::Incoming {
-        let map = self.map;
-        let map2 = map.clone();
-        let future = self.transport
-            .next_incoming()
-            .map(move |upgrade| {
-                let future = upgrade.map_err(map);
-                Box::new(future) as Box<_>
-            })
-            .map_err(map2);
-        Box::new(future)
+        MapErrIncoming {
+            inner: self.transport.next_incoming(),
+            map: Some(self.map),
+        }
+    }
+}
+
+/// Listening stream for `MapErr`.
+pub struct MapErrListener<T, F>
+where T: Transport {
+    inner: T::Listener,
+    map: F,
+}
+
+impl<T, F> Stream for MapErrListener<T, F>
+where T: Transport,
+    F: FnOnce(IoError) -> IoError + Clone,
+{
+    type Item = MapErrListenerUpgrade<T, F>;
+    type Error = IoError;
+
+    #[inline]
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.inner.poll() {
+            Ok(Async::Ready(Some(value))) => {
+                let map = self.map.clone();
+                let value = MapErrListenerUpgrade {
+                    inner: value,
+                    map: Some(map),
+                };
+                Ok(Async::Ready(Some(value)))
+            },
+            Ok(Async::Ready(None)) => Ok(Async::Ready(None)),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(err) => {
+                let map = self.map.clone();
+                Err(map(err))
+            }
+        }
+    }
+}
+
+/// Listening upgrade future for `MapErr`.
+pub struct MapErrListenerUpgrade<T, F>
+where T: Transport {
+    inner: T::ListenerUpgrade,
+    map: Option<F>,
+}
+
+impl<T, F> Future for MapErrListenerUpgrade<T, F>
+where T: Transport,
+    F: FnOnce(IoError) -> IoError,
+{
+    type Item = (T::Output, T::MultiaddrFuture);
+    type Error = IoError;
+
+    #[inline]
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self.inner.poll() {
```

Comment:
can't we stick to the [future-map-err implementation style](https://docs.rs/futures/0.1.13/src/futures/future/map_err.rs.html#30-34) for this one?

```rust

        let e = match self.inner.poll() {
            Ok(Async::NotReady) => return Ok(Async::NotReady),
            other => other
        };
        e.map_err(self.map.take().expect("poll() called again after error"))
```